File: src/Butil/Bit.Butil/Publics/EyeDropper.cs
using System.Threading.Tasks;
using Microsoft.JSInterop;

namespace Bit.Butil;

/// <summary>
/// Wraps the <see href="https://developer.mozilla.org/en-US/docs/Web/API/EyeDropper">EyeDropper API</see>.
/// </summary>
public class EyeDropper(IJSRuntime js)
{
    /// <summary>True when the runtime exposes <c>window.EyeDropper</c>.</summary>
    public ValueTask<bool> IsSupported() => js.Invoke<bool>("BitButil.eyeDropper.isSupported");

    /// <summary>
    /// Opens the eyedropper and returns the picked sRGB color in hex form (e.g. <c>#1f2937</c>).
    /// Returns null when the user cancels or the runtime can't show the picker.
    /// </summary>
    /// <remarks>
    /// Must be called from a user-gesture handler - the browser will reject the request otherwise.
    /// </remarks>
    public ValueTask<string?> Open() => js.Invoke<string?>("BitButil.eyeDropper.open");
}

File: src/Butil/Bit.Butil/Scripts/eyeDropper.ts
var BitButil = BitButil || {};

(function (butil: any) {
    butil.eyeDropper = {
        isSupported() { return 'EyeDropper' in window; },
        async open() {
            const W = window as any;
            if (typeof W.EyeDropper !== 'function') return null;
            try {
                const dropper = new W.EyeDropper();
                const result = await dropper.open();
                return result?.sRGBHex ?? null;
            } catch {
                // User canceled or the call wasn't tied to a user gesture.
                return null;
            }
        }
    };
}(BitButil));
